The default HTTP client can't handle non-JSON response bodies

**Description**

I just went through a bit of painful debugging because I couldn't figure out why my indexing POST requests were failing in production. This is the error message I got:

> MeiliSearch\Exceptions\FailedJsonDecodingException
> Decoding json payload failed: "Syntax error".
> Caused by: JsonException 

The error came from `MeiliSearch\Http\Client::parseResponse()` which uses `MeiliSearch\Http\Client::jsonDecode()`.

Now the actual problem was that in production, the MeiliSearch instance is sitting behind an nginx reverse proxy and the POST request was hitting the `max_client_body_size`, so nginx returned a 413 error with the default HTML error page.

This was difficult to debug since it only happened in production (I use the docker image for testing, so no reverse proxy in the test environment).

**Expected behavior**

The `Client::parseResponse` method should be able to handle non-JSON responses more gracefully so that a failing request in production produces a helpful error message. The client could either check the `Content-Type` header of the response and only parse it as JSON if appropriate. Or catch the exception thrown by `jsonDecode` and throw a more helpful exception with the HTTP response code.

**Current behavior**

If the MeiliSearch server returns a non-JSON response, you get a `FailedJsonDecodingException` with no indication regarding the actual response code and content.

**Environment**

- OS: Ubuntu 20.04
- MeiliSearch version: v.0.22.0
- meilisearch-php version: v0.19.1

--- a/src/Exceptions/InvalidResponseBodyException.php
+++ b/src/Exceptions/InvalidResponseBodyException.php
@@ -1,0 +1,48 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MeiliSearch\Exceptions;
+
+use Exception;
+use Psr\Http\Message\ResponseInterface;
+
+class InvalidResponseBodyException extends Exception
+{
+    public $httpStatus = 0;
+    public $httpBody = null;
+    public $message = null;
+
+    public function __construct(ResponseInterface $response, $httpBody, $previous = null)
+    {
+        $this->httpStatus = $response->getStatusCode();
+        $this->httpBody = $httpBody;
+        $this->message = $this->getMessageFromHttpBody() ?? $response->getReasonPhrase();
+
+        parent::__construct($this->message, $this->httpStatus, $previous);
+    }
+
+    public function __toString()
+    {
+        $base = 'MeiliSearch InvalidResponseBodyException: Http Status: '.$this->httpStatus;
+
+        if ($this->message) {
+            $base .= ' - Message: '.$this->message;
+        }
+
+        return $base;
+    }
+
+    public function getMessageFromHttpBody(): ?string
+    {
+        if (null !== $this->httpBody) {
+            $rawText = strip_tags($this->httpBody);
+
+            if (!ctype_space($rawText)) {
+                return substr(trim($rawText), 0, 100);
+            }
+        }
+
+        return null;
+    }
+}

--- a/src/Http/Client.php
+++ b/src/Http/Client.php
@@ -13,6 +13,7 @@ use MeiliSearch\Exceptions\ApiException;
 use MeiliSearch\Exceptions\CommunicationException;
 use MeiliSearch\Exceptions\FailedJsonDecodingException;
 use MeiliSearch\Exceptions\FailedJsonEncodingException;
+use MeiliSearch\Exceptions\InvalidResponseBodyException;
 use Psr\Http\Client\ClientExceptionInterface;
 use Psr\Http\Client\ClientInterface;
 use Psr\Http\Client\NetworkExceptionInterface;
@@ -191,6 +192,10 @@ class Client implements Http
     {
         if (204 === $response->getStatusCode()) {
             return null;
+        }
+
+        if (!\in_array('application/json', $response->getHeader('content-type'), true)) {
+            throw new InvalidResponseBodyException($response, $response->getBody()->getContents());
         }
 
         if ($response->getStatusCode() >= 300) {

--- a/tests/Exceptions/InvalidResponseBodyExceptionTest.php
+++ b/tests/Exceptions/InvalidResponseBodyExceptionTest.php
@@ -1,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\Exceptions;
+
+use Http\Discovery\Psr17FactoryDiscovery;
+use MeiliSearch\Exceptions\InvalidResponseBodyException;
+use Tests\TestCase;
+
+final class InvalidResponseBodyExceptionTest extends TestCase
+{
+    public function testException(): void
+    {
+        $httpBodyExample = '<b>Gateway Timeout</b>';
+        $statusCode = 504;
+
+        try {
+            $streamFactory = Psr17FactoryDiscovery::findStreamFactory();
+            $responseBodyStream = $streamFactory->createStream($httpBodyExample);
+
+            $responseFactory = Psr17FactoryDiscovery::findResponseFactory();
+            $response = $responseFactory->createResponse($statusCode)->withBody($responseBodyStream);
+
+            throw new InvalidResponseBodyException($response, $httpBodyExample);
+        } catch (InvalidResponseBodyException $invalidResponseBodyException) {
+            $this->assertEquals($statusCode, $invalidResponseBodyException->httpStatus);
+            $this->assertEquals('Gateway Timeout', $invalidResponseBodyException->message);
+
+            $expectedExceptionToString = "MeiliSearch InvalidResponseBodyException: Http Status: {$statusCode} - Message: Gateway Timeout";
+            $this->assertEquals($expectedExceptionToString, (string) $invalidResponseBodyException);
+        }
+    }
+
+    public function testExceptionWithNoHttpBody(): void
+    {
+        $statusCode = 504;
+        $responseFactory = Psr17FactoryDiscovery::findResponseFactory();
+        $response = $responseFactory->createResponse($statusCode);
+
+        try {
+            throw new InvalidResponseBodyException($response, null);
+        } catch (InvalidResponseBodyException $invalidResponseBodyException) {
+            $this->assertEquals($statusCode, $invalidResponseBodyException->httpStatus);
+            $this->assertEquals($response->getReasonPhrase(), $invalidResponseBodyException->message);
+
+            $expectedExceptionToString = "MeiliSearch InvalidResponseBodyException: Http Status: {$statusCode} - Message: {$response->getReasonPhrase()}";
+            $this->assertEquals($expectedExceptionToString, (string) $invalidResponseBodyException);
+        }
+    }
+}

--- a/tests/Http/ClientTest.php
+++ b/tests/Http/ClientTest.php
@@ -7,6 +7,7 @@ namespace Tests\Http;
 use MeiliSearch\Exceptions\ApiException;
 use MeiliSearch\Exceptions\FailedJsonDecodingException;
 use MeiliSearch\Exceptions\FailedJsonEncodingException;
+use MeiliSearch\Exceptions\InvalidResponseBodyException;
 use MeiliSearch\Http\Client;
 use PHPUnit\Framework\MockObject\MockObject;
 use PHPUnit\Framework\TestCase;
@@ -128,6 +129,18 @@ class ClientTest extends TestCase
         $client->post('/', '');
     }
 
+    public function testInvalidResponseContentTypeThrowsException(): void
+    {
+        $httpClient = $this->createHttpClientMock(200, '<b>not json</b>', 'text/html');
+
+        $client = new Client('https://localhost', null, $httpClient);
+
+        $this->expectException(InvalidResponseBodyException::class);
+        $this->expectExceptionMessage('not json');
+
+        $client->get('/');
+    }
+
     public function provideStatusCodes(): iterable
     {
         yield [200];
@@ -138,7 +151,7 @@ class ClientTest extends TestCase
     /**
      * @return ClientInterface|MockObject
      */
-    private function createHttpClientMock(int $status = 200, string $content = '{')
+    private function createHttpClientMock(int $status = 200, string $content = '{', string $contentType = 'application/json')
     {
         $stream = $this->createMock(StreamInterface::class);
         $stream->expects(self::once())
@@ -149,6 +162,10 @@ class ClientTest extends TestCase
         $response->expects(self::any())
             ->method('getStatusCode')
             ->willReturn($status);
+        $response->expects(self::any())
+            ->method('getHeader')
+            ->with('content-type')
+            ->willReturn([$contentType]);
         $response->expects(self::once())
             ->method('getBody')
             ->willReturn($stream);

